plugins/pluginmanagers/lazy.nix

```
@@ -1,222 +1,288 @@
+                # `dependencies`. Also use `types.eitherRecursive` instead of
+                # `types.either` here, as using just `types.either` also leads
+                # to stack overflow. 
+                dependencies = mkNullOrOption (eitherRecursive lazyPluginType lazyPluginsListType) ''
+                  A list of plugin names or plugin specs that should be
+                  loaded when the plugin loads. Dependencies are always
+                  lazy-loaded unless specified otherwise. When specifying a
+                  name, make sure the plugin spec has been defined somewhere
+                  else. This can also be a single string such as for a short
+                    plugin url (See: https://lazy.folke.io/spec).
+                '';
+
+                init = mkNullOrLuaFn "init functions are always executed during startup";
+
+                config = mkNullOrStrLuaFnOr (enum [ true ]) ''
+                  config is executed when the plugin loads.
+
+                  The default implementation will automatically run require(MAIN).setup(opts).
+                  Lazy uses several heuristics to determine the plugin's MAIN module automatically based on the plugin's name.
+                  See also opts. To use the default implementation without opts set config to true.
+                '';
+
+                main = mkNullOrOption str ''
+                  You can specify the main module to use for config() and opts(),
+                  in case it can not be determined automatically. See config()
+                '';
+
+                submodules = defaultNullOpts.mkBool true ''
+                  When false, git submodules will not be fetched.
+                  Defaults to true
+                '';
+
+                event =
+
+                  mkNullOrOption (maybeRaw (
+                    either str (listOf str)
+                  )) "Lazy-load on event. Events can be specified as BufEnter or with a pattern like BufEnter *.lua";
+
+                cmd = mkNullOrOption (maybeRaw (either str (listOf str))) "Lazy-load on command";
+
+                ft = mkNullOrOption (maybeRaw (either str (listOf str))) ''
+                  Lazy-load plugin on filetype e.g.:
+
+                  plugins.lazy.plugins = with pkgs.vimPlugins; [
+                    {
+                      pkg = neorg;
+                      # Only load plugin on "norg" filetyes
+                      ft = "norg";
+                      opts = {
+                        load."['core.defaults']" = [ ];
+                      };
+                    }
+                  ];
+                '';
+
+                keys = mkNullOrOption (maybeRaw (either str (listOf str))) "Lazy-load on key mapping";
+
+                module = mkNullOrOption (enum [ false ]) ''
+                  Do not automatically load this Lua module when it's required somewhere.
+                '';
+
+                priority = mkNullOrOption number ''
```

We should usually prefer the defaultNullHelpers for options that'll (eventually) be going through `toLuaObject`.

This is also how we usually document "plugin defaults"

```suggestion
                priority = defaultNullOpts.mkNum 50 ''
```

If you need more control, the following is equivalent:

```nix
priority = defaultNullOpts.mkNum' {
  pluginDefault = 50;
  description = ''
    Only useful for start plugins i.e.
    `plugins.lazy.plugins.<plugin>.lazy = false` to force loading
    certain plugins first.

    It is recommended to set this to a high number for colorschemes.
  '';
}
```

---

defaultNullOpts will usually wrap `maybeRaw` support around the type, allowing users to configure either the declared type _or_ `__raw` lua code.

`pluginDefault` is actually optional, if we don't know one. But it is required when using the "short form" version of the helper functions (the one without the `'`).

---

(When it's a custom option that doesn't map directly to upstream, then we maybe don't want `defaultNullOpts`)